Cli generator name option, replaces 'language' options in CLI and Maven Plugin

jimschubert opened the pull request:
### PR checklist

- [ ] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ ] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [ ] Filed the PR against the correct branch: Default: `master`.
- [ ] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

The language option (--lang and -l in CLI, language in maven plugin) has
drifted from the original intention of defining a top-level language
implementation with variance on library. Generators began to encode
generator type into language (e.g. erlang-client), reused language as
framework-only (e.g. scalatra), or embedded language and framework (e.g.
typescript-angular).

With the 3.0.0 release of openapi-generator, we've moved to a
standardization of these names, which means they no longer fit into the
singular concept of a "language".

We've discussed plans to provide a better user experience around the
CLI, breaking apart the generator lookup into the variations that make
up a generator:

* language
* type
* framework
* version?

To do this, we must first deprecate language and warn users that what
they're currently using (e.g. scalatra, typescript-angular) is being
replaced by generatorName and that language may mean something else in a
later version of the generator.

For discussion, see #39.

Example usage of the maven plugin is explained in the commit message for e97736f. You must first run `mvn clean package install` to stage the 3.0.0-SNAPSHOT from this branch into your local repository. Then, navigate to `modules/openapi-generator-maven-plugin/examples` and run:

```
mvn -f non-java.xml compile
```

This will create an aspnetcore generated example under `target/generated-sources/`.

Warning messages introduced in this PR will look like (maven plugin output):

```
$ mvn -f non-java.xml compile
[INFO] Scanning for projects...
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building sample-project 1.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- openapi-generator-maven-plugin:3.0.0-SNAPSHOT:generate (default) @ sample-project ---
[WARNING] The 'language' option is deprecated and may reference language names only in the next major release (4.0). Please use 'generatorName' instead.
[WARNING] Output directory does not exist, or is inaccessible. No file (.openapi-generator-ignore) will be evaluated.
```

jmini commented:
This is nice, looks good to me!

I think in the CLI there is also the option to output all available langs:
```
openapi-generator-cli langs
```
Or something like that, this should be changed to `generators` (with backward compatibility => we keep `langs` in parallel during one release)

---
There are also error message, when an invalid lang is selected, that says "available languages" => should be changed to "available generators"

---

We also need to update our migration guide:
https://github.com/OpenAPITools/openapi-generator/blob/master/docs/migration-from-swagger-codegen.md

---

I thought you wanted also to fail (have an error) if both "language" and "generator name" are set. Maybe I missed it from your code.

---

If I remember well you also mention something about supporting old generator names during one release to ease the mig.

Maybe we can add this map somewhere:

```
Map<String,String> nameMigrationMap = new HashMap<>();
nameMigrationMap.put("akka-scala","scala-akka");
nameMigrationMap.put("scala", "scala-httpclient");
nameMigrationMap.put("jaxrs", "jaxrs-jersey");
nameMigrationMap.put("qt5cpp", "cpp-qt5");
nameMigrationMap.put("cpprest","cpp-restsdk");
nameMigrationMap.put("tizen", "cpp-tizen");
nameMigrationMap.put("sinatra","ruby-sinatra");
nameMigrationMap.put("swift", "swift2-deprecated");
nameMigrationMap.put("lumen", "php-lumen");
nameMigrationMap.put("slim","php-slim");
nameMigrationMap.put("ze-ph", "php-ze-ph");
nameMigrationMap.put("nancyfx","csharp-nancyfx");
```

And do a if `map.containsKey()` then output a warning and use the new new instead of the old. We can support this for both "language" and "generator-name" options.

jmini commented:
> question: I've left "language" as a required option for the CLI and maven plugin, but I'm wondering if we should go ahead and make both language and generatorName optional from the start? 

I think both optional is good and error if both are set or if both are missing.

jimschubert commented:
@jmini thanks for doing such a thorough evaluation. I hadn't even considered the `langs` command!

I'll make the improvements you've listed above. I think the `nameMigrationMap` check makes the most sense in the new `setGeneratorName` setter. Since I pass `setLang` to this method, it would be a single point in code to manage the migration mapping.

I'll change the title of the PR to indicate it's a work in progress. Should we add a `WIP` label as well?

jmini commented on modules/openapi-generator-maven-plugin/src/main/java/org/openapitools/codegen/plugin/CodeGenMojo.java:
```diff
@@ -361,7 +372,17 @@ public void execute() throws MojoExecutionException {
+            configurator.setGeneratorName(language);
+        } else {
+            LOGGER.error("A generator name (generatorName) is required.");
+            System.exit(1);
```
I think it is better to throw a `MojoExecutionException` instead of System exist (maven will terminate with an appropriate error message and can still do some other stuff)

jimschubert replied:
Thanks for catching this. I was just blindly copying what I added to the CLI code, and I can't believe I left System.exit here. 💃

jmini commented on modules/openapi-generator-maven-plugin/src/main/java/org/openapitools/codegen/plugin/CodeGenMojo.java:
```diff
@@ -361,7 +372,17 @@ public void execute() throws MojoExecutionException {
 
-        configurator.setLang(language);
+        // TODO: After 3.0.0 release (maybe for 3.1.0): Fully deprecate lang.
+        if (isNotEmpty(generatorName)) {
```
Right after the if, I would add something:

```
if (isNotEmpty(language))
    LOGGER.warn("The 'language' option is deprecated and was replaced by 'generatorName'. Both can not be set together");
    throw new MojoExecutionException("Illegal configuration: 'language' and  'generatorName' can not be set both, remove 'language' from your configuration");
}
```

jimschubert commented:
@jmini all great feedback. Thanks!

Do you think it's necessary to update the migration doc? It's not required to make the changes to the language switch in order to migrate, and it will want the user. I guess now that it's not required to change the language values, either, that it makes sense to list for consistency.

jmini commented:
@jimschubert: I have named it "migration guide" but this is also a "what are the main differences between Swagger-Codegen and OpenAPI Generator".

After **New maven coordinates** section I would create 2 new sections:

* Changes in the CLI
* Changes in Maven Plugin

In each new section I would summarize what have changed with one or two sentences.

In **New generators names** section, I would add one sentence to say that OpenAPI Generator is tolerant with old names during one release.

jimschubert commented:
@jmini Thanks again for the second pair of eyes.

I had originally left all `-l` instances in scripts because the option is still valid, but on second thought I wanted these to be "current" and not produce a warning by default on initial release. Because there were about 500 locations that all had to be verified individually and changed manually, I've broken changes down into many separate commits to make them easier to scan.

I was wondering, what determines that docs should exist under `/docs` rather than in the wiki? It looked like you created or edited most of those files, and some of the information seems duplicated or like there are empty placeholders on the wiki: 

* https://github.com/OpenAPITools/openapi-generator/blob/master/docs/integration.md vs https://github.com/OpenAPITools/openapi-generator/wiki/Workflow-integration
* https://github.com/OpenAPITools/openapi-generator/blob/master/docs/customization.md vs https://github.com/OpenAPITools/openapi-generator/wiki/Customizing-client-code-generation ) and others)

Are we intending to move those under the folder to the wiki, or are we planning to deprecate the wiki in favor of source control docs?

jmini commented:
> It looked like you created or edited most of those files

@wing328 extracted those files from the README with c32ef12c8ef2014d909372752aac9bbfd7fbf4e7

I just moved them into a `docs/` folder with 3d7eaf96fcf13b23eb9165ccbfe3054c94cc6dc6

---

**Yes** we we have planned to move more content into the repository (and remove them from the wiki). 

Multiple reasons for that:

* Documentation belongs to software change, have a look at this PR, it is great to be able to discuss what needs to be documented.
* GitHub Wiki does not support PR workflow, which is a great way to do suggestions, reviews and improvements
* Wiki are not forked in GitHub forks, you need to migrate them separately (they are git repositories, but not well integrated in GitHub UI).
* Software documentation follows the software branches. Imagine we works on 4.0 and do some bugfix release for 3.0 on a separated `3.x-maintenance` branch. With a GitHub wiki you just have a `master` branch, you can not support both versions in parallel.
* On the long term the plan is to generate the documentation inside a static website. For that we might need some artefacts that comes from the code itself (like for example the help in the command line that you have updated by hand). In the docs page there should be a placeholder that some content will be injected, and the static-website build tool should put the correct values in the built html manual. When we have this it make even more sense to single source everything in one repo and not in two.

I would say that we should migrate the content of the wiki into some files in the `docs/` folder. This can be a step-by-step process. I you spot something outdated in the wiki that have already its counterpart in the `docs` just delete it and replace it with a link to the new page.

jimschubert commented:
@jmini I fully agree with moving things to `/docs` for the reasons you posted above. Also, docs in the repository support all of github's templating (asciidoc, csv rendering, etc).

I think we should migrate everything from the wiki, as I find it confusing that there are multiple places for all of these details and I'm not easily able to do a search across the project from the wiki. I could fork the wiki into a subdirectory, but that wouldn't share a commit with this PR.

wing328 commented:
The FAQ (wiki) that I originally created intends to be community driven FAQ as anyone can update it easily without a PR. (we did see contributions from the community to add more FAQs over time)

The Q&A (mainly for the launch of openapi-generator)  is not something I want people to update without approval/review as it represents our viewpoints.

I agree it may confuse users and happy to see consolidations or other approaches to make our doc more useful and easily discoverable by our users.

jmini commented:
This discussion about `docs/` has nothing to do with this PR, I have opened #103 (it will be easier to find it later).

jimschubert commented:
@jmini I think the `/docs` discussion is relevant here, as it has lots of `-l` occurrences that would need to be changed once this is merged. I don't think the `/docs` vs wiki discussion has to be solved here, just whether those pages are intended to be kept up-to-date or if the `/docs` were only a placeholder as we migrated the wiki.  What I didn't want to happen is for me to move content from `/docs` to the wiki or vice versa as a follow up.

Sorry if I didn't clarify the reasoning for my question previously.

jmini commented:
It is OK. Better more text to clarify than misunderstanding. I have updated my comment to say that the "docs vs wiki discussion" was moved to an other issue.

It is great that you have updated the docs, and the `bin/` scripts.

Tell us when you consider this ready for test.

jimschubert commented:
@jmini this is ready to test. Sorry, I thought I had included that in my comment after the last push.

jmini commented:
@jimschubert can you solve the conflicts created by #124 so that we can merge this?

During the merge, for each conflict inside the `bin/` folder you can take the `master` version and then redo a search and replace `-l` => `-g`.

jimschubert commented:
@jmini @wing328 I've fixed these merge conflicts. I've also added an exit code of 1 if the CLI is called without args, to indicate failure. I didn't test whether the airlift framework also does this on `help`, which is common behavior… but I think that's not important unless we change the default command to `help`.
